### PR TITLE
Various Renderer Crash Improvements

### LINKED ISSUE #5879: [ACRA 2.9] WebViewRenderer Crash - Async NullPointerException
Between the two lines, `cardFrame` is null, this is accessed by an AsyncTask on `blockControls`.

https://github.com/ankidroid/Anki-Android/blob/6e19f68fdc8088352e3ca8785d080e2a5f0a4276/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java#L1615-L1650

----

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/bbbb5b62-7867-412a-84dd-0fba1c518ff9

```
java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.FrameLayout.setEnabled(boolean)' on a null object reference
at com.ichi2.anki.AbstractFlashcardViewer.blockControls(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.AbstractFlashcardViewer.access$1300(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.AbstractFlashcardViewer$9.onPreExecute(AbstractFlashcardViewer.java:1)
at com.ichi2.async.DeckTask$TaskListener.onPreExecute(DeckTask.java:1)
at com.ichi2.async.DeckTask.onPreExecute(DeckTask.java:2)
at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:708)
at android.os.AsyncTask.execute(AsyncTask.java:655)
at com.ichi2.async.DeckTask.launchDeckTask(DeckTask.java:2)
at com.ichi2.anki.AbstractFlashcardViewer.answerCard(AbstractFlashcardViewer.java:18)
at com.ichi2.anki.AbstractFlashcardViewer$5.onClick(AbstractFlashcardViewer.java:13)
at android.view.View.performClick(View.java:7870)
at com.ichi2.anki.AbstractFlashcardViewer$16.run(AbstractFlashcardViewer.java:2)
at android.os.Handler.handleCallback(Handler.java:883)
at android.os.Handler.dispatchMessage(Handler.java:100)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:7811)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
```

### PR DESCRIPTION
**Pull request by david-allison-1**
##  Purpose / Description

* Fixes Gestures being disabled after renderer is regenerated
* Fixes various race conditions when the render ends
* Fixes a dialog box for a renderer error being cancelable, not ending the activity as expected.

## Fixes
Fixes #5879

## Approach

* Gestures: Insert the view at position 0, rather than position 3 (touches appear to be from lowest to highest).
* Fixes race condition accessing `mCardFrame` via atomic variable replacement
* Fixes race condition accessing `mCardFrameParent` inside `onRenderProcessGone` via extracting the constant reference
* Fixes race condition accessing `mCard` via reader-writer lock
* Made error dialog for render process failure non-cancelable

## How Has This Been Tested?

On my device. I can no longer cause a crash due to mashing the answer buttons/gestures due to a race condition.

There is still an unsolved crash on the first instance of generating the renderer. This previously existed in the branch

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

## Review Notes

* Happy to split this into two PRs (one handling `mCard`, one for the remainder of the fixes)

### CODE REVIEW

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5276672&bid=17279787)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -2242,30 +2303,30 @@ private void addAnswerSounds(String answer) {
     }
 
 
-    private void updateCard(String content) {
+    private void updateCard(final String newContent) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: Codacy Issue found: [The method 'updateCard(String)' has an NPath complexity of 7680](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5276672)

**Comment by david-allison-1** [comment]
> These look fine to me - thanks for separating them into multiple comments - would have been quite difficult to read through without that but was easy with it.
> 
> Do you feel like these are ready for an alpha build? If they stabilize there I would assume a pick to 2.9.x would be good as well as this is all related to #5780 yes?

Feels good to go into alpha to me. All seems reasonable. 

I'm still unsure about the `System.gc` call, I don't think it's necessary, or helps but will keep it in there until someone works out a fix for the remaining issue for #5780

**Comment by mikehardy** [comment]
2.10alpha59 will have this as soon as google's done chewing on it - could you open a separate issue for the remaining bit of #5780? If anything we'll target a 2.9.7 with it but I want to get 2.9.6 out the door, it's pretty big now both in amount it will help people and amount of change so I don't want more delay if it can be avoided. After this marinates in alpha a day (so at least @sudomain or @Anthropos888 can maybe give the alpha a check :-) ) I'll pick to another 2.9.6 beta then hopefully release the thing

**Pull request merged**